core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala

```diff
@@ -376,18 +374,13 @@ private[netty] class NettyRpcEnv(
 
     def setError(e: Throwable): Unit = {
       error = e
-      source.close()
     }
 
     override def read(dst: ByteBuffer): Int = {
       Try(source.read(dst)) match {
+        case _ if error != null => throw error
```

Comment:
I think it is better to also add a short comment here. This bug is subtle and no test against it now. Just from this code, it is hard to know why we check error even success.

Reply:
I added a pair of comments to explain the flow of calls involving `setError()` and pipe closes.